### PR TITLE
Synchronous join protocol

### PR DESCRIPTION
**Pull request by jumaffre**
This PR adds support for synchronous join protocol.

Before, the join RPC would return immediately, i.e. before knowing whether the join protocol actually succeeded. Now, if `"is_sync" = true` is passed via the `"joinNetwork"` RPC, the command will return only when the leader has replied to the joining node. 

The end-to-end tests have been updated accordingly (defaults to synchronous join but still possible to use the asynchronous protocol via the `--asynchronous-join` flag)

**Also**
- I have unified the RPC context for both non-forwarded and forwarded commands.
- I have refactored some of `frontend.h` so that the RPC context holds more information (useful for join protocol).
- Since forwarding does not forward the caller certificate (only the caller id), I have marked the only handler we have (member `ACK`) that use the caller certificate explicitly as "non forwardable" (i.e. it has to be run on the leader and cannot be forwarded by a follower).

### CODE REVIEW

File: src/node/nodestate.h [hunk 1]
```diff
@@ -326,13 +326,22 @@ namespace ccf
...
 
       // Create and connect to endpoint
-      auto join_client = rpcsessions.create_client(join_client_cert);
+      auto join_client =
+        rpcsessions.create_client(rpc_ctx, std::move(join_client_cert));
+
+      // If the join network command is synchronous, only reply to the client
+      // when the leader has responded
+      if (args.is_sync)
+        rpc_ctx.is_suspended = true;
```
Comment (jumaffre) [inline src/node/nodestate.h]: This flag used to be called `is_forwarded` but needed renaming as the feature is now more generic. Any better names for this flag, please let me know.

File: src/enclave/enclavetypes.h [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  {
+    //
+    // In parameters (initialised when context is created)
+    //
+
+    // Client session ID
+    const size_t session_id = InvalidSessionId;
+    // Actor type to route to appropriate frontend
+    const ccf::ActorsType actor;
+    // Caller certificate
```
Comment (achamayou) [inline src/enclave/enclavetypes.h]: s/caller/caller_cert and kill the comment?

File: src/enclave/enclavetypes.h [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  static constexpr size_t InvalidSessionId = std::numeric_limits<size_t>::max();
+
+  struct RPCContext
+  {
+    //
+    // In parameters (initialised when context is created)
+    //
+
+    // Client session ID
+    const size_t session_id = InvalidSessionId;
```
Comment (achamayou) [inline src/enclave/enclavetypes.h]: s/session_id/client_session_id

File: src/enclave/enclavetypes.h [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    // Actor type to route to appropriate frontend
+    const ccf::ActorsType actor;
+    // Caller certificate
+    CBuffer caller;
+
+    //
+    // Out parameters (changed during lifetime of context)
+    //
+
+    // If true, the RPC does not reply to the client synchronously
```
Comment (achamayou) [inline src/enclave/enclavetypes.h]: pending?

File: src/enclave/enclavetypes.h [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    bool is_suspended = false;
+    // Packing method used to serialise the RPC object
+    std::optional<jsonrpc::Pack> pack = std::nullopt;
+    // JSON-RPC specific
+    struct json
+    {
+      uint64_t seq_no;
+    };
+    struct json json;
```
Comment (achamayou) [inline src/enclave/enclavetypes.h]: jsonrpc

Comment (achamayou) [reply src/enclave/enclavetypes.h]: or maybe request

File: src/enclave/enclavetypes.h [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    // Only set in the case of a forwarded RPC
+    //
+    struct forwarded
+    {
+      // Initialised when forwarded context is created
+      const size_t session_id;
+      const ccf::NodeId from;
+      const ccf::CallerId caller_id;
+
+      // Changed during lifetime of forwarded context
```
Comment (achamayou) [inline src/enclave/enclavetypes.h]: optional?

File: src/enclave/rpcsessions.h [hunk 7]
```diff
@@ -78,15 +79,15 @@ namespace enclave
...
+    void reply_async(size_t id, const std::vector<uint8_t>& data) override
     {
       std::lock_guard<SpinLock> guard(lock);
 
       auto search = sessions.find(id);
       if (search == sessions.end())
       {
         throw std::logic_error(
-          "reply forwarded for unknown session: " + std::to_string(id));
+          "reply async for unknown session: " + std::to_string(id));
```
Comment (achamayou) [inline src/enclave/rpcsessions.h]: fmt::format("reply async for unknown session: {}", id)

File: src/genesisgen/main.cpp [hunk 8]
```diff
@@ -193,6 +193,12 @@ int main(int argc, char** argv)
   join_rpc->add_option(
     "--join-json", join_json, "Join network RPC as a JSON file", true);
 
+  bool join_sync;
+  join_rpc->add_flag(
```
[the remark this answers is not in the document]
Comment (jumaffre) [reply src/genesisgen/main.cpp]: I have removed this option and the synchronous join protocol is now the only way to perform the join protocol.

**Pull request merged**